File: easy/happy_numbers.clj
(defn in? [seq elm]
  (some #(= elm %) seq))
(defn happy [a]
  (if (< a 10)
    (* a a)
    (+ (* (mod a 10) (mod a 10)) (happy (quot a 10)))))
(defn ishappy [a l]
  (if (= a 1)
    1
    (if (in? l a)
      0
      (ishappy (happy a) (conj l a)))))

(with-open [rdr (clojure.java.io/reader (first *command-line-args*))]
  (doseq [line (line-seq rdr)]
    (prn (ishappy (Integer/parseInt line) []))))

File: easy/happy_numbers.pl
use strict;
use integer;

sub happy {
    my ( $a, $r ) = @_[0], 0;
    while ( $a > 0 ) {
        my $b = $a % 10;
        $r += $b * $b;
        $a /= 10;
    }
    return $r;
}

open( INFILE, "<", $ARGV[0] )
  or die("Cannot open file $ARGV[0] for reading: $!");
while ( my $line = <INFILE> ) {
    chomp $line;
    my %b = ( $line => 1 );
    for ( my $i = 0 ; $i < 127 && $line > 1 ; $i++ ) {
        $line = happy($line);
        $line == 0 if ( $b{$line} );
        $b{$line} = 1;
    }
    printf( "%d\n", $line == 1 );
}
close(INFILE);
